core/src/main/scala/com/eed3si9n/jarjarabrams/Zip.scala

```diff
@@ -1,192 +1,134 @@
-      f: ZipOutputStream => Unit
-  ) = {
-    Using.fileOutputStream(false)(file) { fileOut =>
-      val (zipOut, _) =
-        manifest match {
-          case Some(mf) =>
-            import Attributes.Name.MANIFEST_VERSION
-            val main = mf.getMainAttributes
-            if (!main.containsKey(MANIFEST_VERSION))
-              main.put(MANIFEST_VERSION, "1.0")
-            val os = new JarOutputStream(fileOut)
-            val e = new ZipEntry(JarFile.MANIFEST_NAME)
-            e.setTime(time)
-            os.putNextEntry(e)
-            mf.write(new BufferedOutputStream(os))
-            os.closeEntry()
-            (os, "jar")
-          case None => (new ZipOutputStream(fileOut), "zip")
-        }
-      try {
-        f(zipOut)
-      } finally {
-        zipOut.close
+  // ZIP timestamps have a resolution of 2 seconds.
+  // see http://www.info-zip.org/FAQ.html#limits
+  final private val minimumTimestampIncrement = 2000L
+  // 1980-01-03
+  // Technically it should be 1980-01-01, but Java checks against
+  // the year segment in localtime and ends up capturing timezone,
+  // so to be safe we should consider Jan 3 to be the minimum.
+  //
+  //   $ unzip -l example/byte-buddy-agent.jar | head
+  // Archive:  example/byte-buddy-agent.jar
+  //   Length      Date    Time    Name
+  // ---------  ---------- -----   ----
+  //         0  00-00-1980 04:08   META-INF/
+  //       988  00-00-1980 04:08   META-INF/MANIFEST.MF
+  private final val minimumTimestamp = 315705600L
+
+  def list(inputJar: Path): List[(String, Long)] =
+    Using.jarFile(inputJar) { in =>
+      in.entries.asScala.toList.map { entry =>
+        (entry.getName, entry.getTime)
       }
     }
-  }
-
-  private def unzipStream(
-      from: InputStream,
-      toDirectory: Path,
-      preserveLastModified: Boolean = false
-  ): Set[Path] = {
-    createDirectories(toDirectory)
-    Using.zipInputStream(from)(zipInput => extract(zipInput, toDirectory, preserveLastModified))
-  }
 
-  private def extract(
-      from: ZipInputStream,
-      toDirectory: Path,
-      preserveLastModified: Boolean
-  ) = {
-    val set = new mutable.HashSet[Path]
-    @tailrec def next(): Unit = {
-      val entry = from.getNextEntry
-      if (entry == null) ()
-      else {
-        val name = entry.getName
-        val target = toDirectory.resolve(name)
-        if (entry.isDirectory) createDirectories(target)
-        else {
-          set += target
-          try {
-            Using.fileOutputStream(false)(target)(out => transfer(from, out))
-          } catch {
-            case NonFatal(e) =>
-              throw new RuntimeException(
-                "error extracting Zip entry '" + name + "' to '" + target + "'",
-                e
-              )
+  /**
+   * Treating JAR file as a set of EntryStruct, this implements a
+   * functional processing, intended for in-memory shading.
+   */
+  def flatMap(
```

Comment:
I'm pretty negative on this name. I'm a bit of a flatMap-means-monad guy personally and this is pretty far from.

What about `def maybeTransformEntries` or rewriteOption or something.

Reply:
Yea, I knew it wasn't really `flatMap` but it's a good mental approximation to think of it as such. Renamed to `transformJarFile` in https://github.com/eed3si9n/jarjar-abrams/pull/39/commits/d41bfc7e682966948cea777d4cbe8e9db4ca1b65.